// @kubernetes/client-node/dist/gen/api/apis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIS = exports.HttpError = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./admissionregistrationApi"), exports);
const admissionregistrationApi_1 = require("./admissionregistrationApi");
tslib_1.__exportStar(require("./admissionregistrationV1Api"), exports);
const admissionregistrationV1Api_1 = require("./admissionregistrationV1Api");
tslib_1.__exportStar(require("./apiextensionsApi"), exports);
const apiextensionsApi_1 = require("./apiextensionsApi");
tslib_1.__exportStar(require("./apiextensionsV1Api"), exports);
const apiextensionsV1Api_1 = require("./apiextensionsV1Api");
tslib_1.__exportStar(require("./apiregistrationApi"), exports);
const apiregistrationApi_1 = require("./apiregistrationApi");
tslib_1.__exportStar(require("./apiregistrationV1Api"), exports);
const apiregistrationV1Api_1 = require("./apiregistrationV1Api");
tslib_1.__exportStar(require("./apisApi"), exports);
const apisApi_1 = require("./apisApi");
tslib_1.__exportStar(require("./appsApi"), exports);
const appsApi_1 = require("./appsApi");
tslib_1.__exportStar(require("./appsV1Api"), exports);
const appsV1Api_1 = require("./appsV1Api");
tslib_1.__exportStar(require("./authenticationApi"), exports);
const authenticationApi_1 = require("./authenticationApi");
tslib_1.__exportStar(require("./authenticationV1Api"), exports);
const authenticationV1Api_1 = require("./authenticationV1Api");
tslib_1.__exportStar(require("./authorizationApi"), exports);
const authorizationApi_1 = require("./authorizationApi");
tslib_1.__exportStar(require("./authorizationV1Api"), exports);
const authorizationV1Api_1 = require("./authorizationV1Api");
tslib_1.__exportStar(require("./autoscalingApi"), exports);
const autoscalingApi_1 = require("./autoscalingApi");
tslib_1.__exportStar(require("./autoscalingV1Api"), exports);
const autoscalingV1Api_1 = require("./autoscalingV1Api");
tslib_1.__exportStar(require("./autoscalingV2beta1Api"), exports);
const autoscalingV2beta1Api_1 = require("./autoscalingV2beta1Api");
tslib_1.__exportStar(require("./autoscalingV2beta2Api"), exports);
const autoscalingV2beta2Api_1 = require("./autoscalingV2beta2Api");
tslib_1.__exportStar(require("./batchApi"), exports);
const batchApi_1 = require("./batchApi");
tslib_1.__exportStar(require("./batchV1Api"), exports);
const batchV1Api_1 = require("./batchV1Api");
tslib_1.__exportStar(require("./batchV1beta1Api"), exports);
const batchV1beta1Api_1 = require("./batchV1beta1Api");
tslib_1.__exportStar(require("./certificatesApi"), exports);
const certificatesApi_1 = require("./certificatesApi");
tslib_1.__exportStar(require("./certificatesV1Api"), exports);
const certificatesV1Api_1 = require("./certificatesV1Api");
tslib_1.__exportStar(require("./coordinationApi"), exports);
const coordinationApi_1 = require("./coordinationApi");
tslib_1.__exportStar(require("./coordinationV1Api"), exports);
const coordinationV1Api_1 = require("./coordinationV1Api");
tslib_1.__exportStar(require("./coreApi"), exports);
const coreApi_1 = require("./coreApi");
tslib_1.__exportStar(require("./coreV1Api"), exports);
const coreV1Api_1 = require("./coreV1Api");
tslib_1.__exportStar(require("./customObjectsApi"), exports);
const customObjectsApi_1 = require("./customObjectsApi");
tslib_1.__exportStar(require("./discoveryApi"), exports);
const discoveryApi_1 = require("./discoveryApi");
tslib_1.__exportStar(require("./discoveryV1Api"), exports);
const discoveryV1Api_1 = require("./discoveryV1Api");
tslib_1.__exportStar(require("./discoveryV1beta1Api"), exports);
const discoveryV1beta1Api_1 = require("./discoveryV1beta1Api");
tslib_1.__exportStar(require("./eventsApi"), exports);
const eventsApi_1 = require("./eventsApi");
tslib_1.__exportStar(require("./eventsV1Api"), exports);
const eventsV1Api_1 = require("./eventsV1Api");
tslib_1.__exportStar(require("./eventsV1beta1Api"), exports);
const eventsV1beta1Api_1 = require("./eventsV1beta1Api");
tslib_1.__exportStar(require("./flowcontrolApiserverApi"), exports);
const flowcontrolApiserverApi_1 = require("./flowcontrolApiserverApi");
tslib_1.__exportStar(require("./flowcontrolApiserverV1beta1Api"), exports);
const flowcontrolApiserverV1beta1Api_1 = require("./flowcontrolApiserverV1beta1Api");
tslib_1.__exportStar(require("./internalApiserverApi"), exports);
const internalApiserverApi_1 = require("./internalApiserverApi");
tslib_1.__exportStar(require("./internalApiserverV1alpha1Api"), exports);
const internalApiserverV1alpha1Api_1 = require("./internalApiserverV1alpha1Api");
tslib_1.__exportStar(require("./logsApi"), exports);
const logsApi_1 = require("./logsApi");
tslib_1.__exportStar(require("./networkingApi"), exports);
const networkingApi_1 = require("./networkingApi");
tslib_1.__exportStar(require("./networkingV1Api"), exports);
const networkingV1Api_1 = require("./networkingV1Api");
tslib_1.__exportStar(require("./nodeApi"), exports);
const nodeApi_1 = require("./nodeApi");
tslib_1.__exportStar(require("./nodeV1Api"), exports);
const nodeV1Api_1 = require("./nodeV1Api");
tslib_1.__exportStar(require("./nodeV1alpha1Api"), exports);
const nodeV1alpha1Api_1 = require("./nodeV1alpha1Api");
tslib_1.__exportStar(require("./nodeV1beta1Api"), exports);
const nodeV1beta1Api_1 = require("./nodeV1beta1Api");
tslib_1.__exportStar(require("./openidApi"), exports);
const openidApi_1 = require("./openidApi");
tslib_1.__exportStar(require("./policyApi"), exports);
const policyApi_1 = require("./policyApi");
tslib_1.__exportStar(require("./policyV1Api"), exports);
const policyV1Api_1 = require("./policyV1Api");
tslib_1.__exportStar(require("./policyV1beta1Api"), exports);
const policyV1beta1Api_1 = require("./policyV1beta1Api");
tslib_1.__exportStar(require("./rbacAuthorizationApi"), exports);
const rbacAuthorizationApi_1 = require("./rbacAuthorizationApi");
tslib_1.__exportStar(require("./rbacAuthorizationV1Api"), exports);
const rbacAuthorizationV1Api_1 = require("./rbacAuthorizationV1Api");
tslib_1.__exportStar(require("./rbacAuthorizationV1alpha1Api"), exports);
const rbacAuthorizationV1alpha1Api_1 = require("./rbacAuthorizationV1alpha1Api");
tslib_1.__exportStar(require("./schedulingApi"), exports);
const schedulingApi_1 = require("./schedulingApi");
tslib_1.__exportStar(require("./schedulingV1Api"), exports);
const schedulingV1Api_1 = require("./schedulingV1Api");
tslib_1.__exportStar(require("./schedulingV1alpha1Api"), exports);
const schedulingV1alpha1Api_1 = require("./schedulingV1alpha1Api");
tslib_1.__exportStar(require("./storageApi"), exports);
const storageApi_1 = require("./storageApi");
tslib_1.__exportStar(require("./storageV1Api"), exports);
const storageV1Api_1 = require("./storageV1Api");
tslib_1.__exportStar(require("./storageV1alpha1Api"), exports);
const storageV1alpha1Api_1 = require("./storageV1alpha1Api");
tslib_1.__exportStar(require("./storageV1beta1Api"), exports);
const storageV1beta1Api_1 = require("./storageV1beta1Api");
tslib_1.__exportStar(require("./versionApi"), exports);
const versionApi_1 = require("./versionApi");
tslib_1.__exportStar(require("./wellKnownApi"), exports);
const wellKnownApi_1 = require("./wellKnownApi");
class HttpError extends Error {
    constructor(response, body, statusCode) {
        super('HTTP request failed');
        this.response = response;
        this.body = body;
        this.statusCode = statusCode;
        this.name = 'HttpError';
    }
}
exports.HttpError = HttpError;
exports.APIS = [admissionregistrationApi_1.AdmissionregistrationApi, admissionregistrationV1Api_1.AdmissionregistrationV1Api, apiextensionsApi_1.ApiextensionsApi, apiextensionsV1Api_1.ApiextensionsV1Api, apiregistrationApi_1.ApiregistrationApi, apiregistrationV1Api_1.ApiregistrationV1Api, apisApi_1.ApisApi, appsApi_1.AppsApi, appsV1Api_1.AppsV1Api, authenticationApi_1.AuthenticationApi, authenticationV1Api_1.AuthenticationV1Api, authorizationApi_1.AuthorizationApi, authorizationV1Api_1.AuthorizationV1Api, autoscalingApi_1.AutoscalingApi, autoscalingV1Api_1.AutoscalingV1Api, autoscalingV2beta1Api_1.AutoscalingV2beta1Api, autoscalingV2beta2Api_1.AutoscalingV2beta2Api, batchApi_1.BatchApi, batchV1Api_1.BatchV1Api, batchV1beta1Api_1.BatchV1beta1Api, certificatesApi_1.CertificatesApi, certificatesV1Api_1.CertificatesV1Api, coordinationApi_1.CoordinationApi, coordinationV1Api_1.CoordinationV1Api, coreApi_1.CoreApi, coreV1Api_1.CoreV1Api, customObjectsApi_1.CustomObjectsApi, discoveryApi_1.DiscoveryApi, discoveryV1Api_1.DiscoveryV1Api, discoveryV1beta1Api_1.DiscoveryV1beta1Api, eventsApi_1.EventsApi, eventsV1Api_1.EventsV1Api, eventsV1beta1Api_1.EventsV1beta1Api, flowcontrolApiserverApi_1.FlowcontrolApiserverApi, flowcontrolApiserverV1beta1Api_1.FlowcontrolApiserverV1beta1Api, internalApiserverApi_1.InternalApiserverApi, internalApiserverV1alpha1Api_1.InternalApiserverV1alpha1Api, logsApi_1.LogsApi, networkingApi_1.NetworkingApi, networkingV1Api_1.NetworkingV1Api, nodeApi_1.NodeApi, nodeV1Api_1.NodeV1Api, nodeV1alpha1Api_1.NodeV1alpha1Api, nodeV1beta1Api_1.NodeV1beta1Api, openidApi_1.OpenidApi, policyApi_1.PolicyApi, policyV1Api_1.PolicyV1Api, policyV1beta1Api_1.PolicyV1beta1Api, rbacAuthorizationApi_1.RbacAuthorizationApi, rbacAuthorizationV1Api_1.RbacAuthorizationV1Api, rbacAuthorizationV1alpha1Api_1.RbacAuthorizationV1alpha1Api, schedulingApi_1.SchedulingApi, schedulingV1Api_1.SchedulingV1Api, schedulingV1alpha1Api_1.SchedulingV1alpha1Api, storageApi_1.StorageApi, storageV1Api_1.StorageV1Api, storageV1alpha1Api_1.StorageV1alpha1Api, storageV1beta1Api_1.StorageV1beta1Api, versionApi_1.VersionApi, wellKnownApi_1.WellKnownApi];
