// @kubernetes/client-node/dist/gen/model/v1StorageClass.js
"use strict";
/**
 * Kubernetes
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: v1.22.2
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1StorageClass = void 0;
/**
* StorageClass describes the parameters for a class of storage for which PersistentVolumes can be dynamically provisioned.  StorageClasses are non-namespaced; the name of the storage class according to etcd is in ObjectMeta.Name.
*/
class V1StorageClass {
    static getAttributeTypeMap() {
        return V1StorageClass.attributeTypeMap;
    }
}
exports.V1StorageClass = V1StorageClass;
V1StorageClass.discriminator = undefined;
V1StorageClass.attributeTypeMap = [
    {
        "name": "allowVolumeExpansion",
        "baseName": "allowVolumeExpansion",
        "type": "boolean"
    },
    {
        "name": "allowedTopologies",
        "baseName": "allowedTopologies",
        "type": "Array<V1TopologySelectorTerm>"
    },
    {
        "name": "apiVersion",
        "baseName": "apiVersion",
        "type": "string"
    },
    {
        "name": "kind",
        "baseName": "kind",
        "type": "string"
    },
    {
        "name": "metadata",
        "baseName": "metadata",
        "type": "V1ObjectMeta"
    },
    {
        "name": "mountOptions",
        "baseName": "mountOptions",
        "type": "Array<string>"
    },
    {
        "name": "parameters",
        "baseName": "parameters",
        "type": "{ [key: string]: string; }"
    },
    {
        "name": "provisioner",
        "baseName": "provisioner",
        "type": "string"
    },
    {
        "name": "reclaimPolicy",
        "baseName": "reclaimPolicy",
        "type": "string"
    },
    {
        "name": "volumeBindingMode",
        "baseName": "volumeBindingMode",
        "type": "string"
    }
];
